src/Strategy/Python/Poetry.hs

```diff
@@ -0,0 +1,181 @@
+      _ <- logIgnoredDeps pyproject (Just poetryLockProject)
+      context "Building dependency graph from pyproject.toml and poetry.lock" $ pure $ buildGraphWithLock poetryLockProject pyproject
+    Nothing -> context "Building dependency graph from only pyproject.toml" $ pure $ Graphing.fromList $ pyProjectDeps pyproject
+
+-- | Build graph of `Dependency` only from `PyProject` and `PoetryLock`.
+buildGraphWithLock :: PoetryLock -> PyProject -> Graphing Dependency
+buildGraphWithLock lockProject poetryProject = promoteToDirect isDirect graph
+  where
+    -- Dependencies in `poetry.lock` are direct if they're specified in `pyproject.toml`.
+    isDirect :: Dependency -> Bool
+    isDirect dep = case pyprojectPoetry poetryProject of
+      Nothing -> False
+      Just _ -> any (\n -> dependencyName n == dependencyName dep) (pyProjectDeps poetryProject)
+
+    lowerCasedPkgName :: PackageName -> PackageName
+    lowerCasedPkgName name = PackageName <$> toLower $ unPackageName name
+
+    graph = gmap pkgNameToDependency (buildPackageNameGraph $ poetryLockPackages lockProject)
+    mapOfDependency = toMap $ poetryLockPackages lockProject
+
+    -- Pip packages are [case insensitive](https://www.python.org/dev/peps/pep-0508/#id21), but poetry.lock may use
+    -- non-canonical name for reference. Try to lookup with provided casing, otherwise fallback to lower casing.
+    --
+    -- Poetry.lock when referencing deep dependencies uses non canonical names.
+    --
+    --  ```toml
+    --  [package.dependencies]
+    --  MarkupSafe = ">=2.0"
+    --  ....
+    --
+    -- [[package]]
+    -- name = "markupsafe"
+    -- version = "2.0.1"
+    -- ...
+    -- ```
+    --
+    -- TODO: Better approach to handle casing scenarios with poetry.lock file.
+    pkgNameToDependency name =
+      fromMaybe
+        ( Dependency
+            { dependencyType = PipType
+            , dependencyName = unPackageName name
+            , dependencyVersion = Nothing
+            , dependencyLocations = []
+            , dependencyEnvironments = []
+            , dependencyTags = Map.empty
+            }
+        )
+        $ Map.lookup name mapOfDependency
+          <|> Map.lookup (lowerCasedPkgName name) mapOfDependency
+
+-- | Builds the Package Name Graph.
+-- In python, package names are unique, and only single version of a package can be used.
+buildPackageNameGraph :: [PoetryLockPackage] -> Graphing PackageName
+buildPackageNameGraph pkgs = foldr deep edges pkgsNoDeps
+  where
+    pkgsNoDeps :: [PackageName]
+    pkgsNoDeps = poetryLockPackageName <$> filter (null . poetryLockPackageDependencies) pkgs
+
+    depsWithEdges :: [PoetryLockPackage]
+    depsWithEdges = filter (not . null . poetryLockPackageDependencies) pkgs
+
+    edgeOf :: PoetryLockPackage -> [(PackageName, PackageName)]
+    edgeOf p = (poetryLockPackageName p,) <$> (PackageName <$> Map.keys (poetryLockPackageDependencies p))
```

Comment:
This function struck me as odd, and I couldn't figure out why until I realized that the functor you're using in both of those `<$>`'s is just `List`, and this is just a `map f . map g`, which is the same as `map f . g`.  That's actually a functor law.  `fmap` for lists is just `map`.  

If you rewrite these to use `map` instead, hlint will complain that you should only use map once (otherwise you iterate twice).

We can simplify this a little:
```suggestion
    edgeOf p = map tuplify . Map.keys $ poetryLockPackageDependencies p
      where
        tuplify x = (poetryLockPackageName p, PackageName x)
```

Reply:
I've updated the code. but curious on, why would this would iterate twice? 

Also using map for list - I'm perplexed on why we have `map` and `fmap` when they fmap works for any type class.

Reply:
> I've updated the code. but curious on, why would this would iterate twice?

Let's dive in a little deeper.  Note that these steps are _exactly_ the steps I took on my side to rewrite the code.

```haskell
-- Original
edgeOf p = (poetryLockPackageName p,) <$> (PackageName <$> Map.keys (poetryLockPackageDependencies p))

-- We can rewrite this to use `fmap` instead of `<$>`, which is exactly equivalent
-- Remember that `a <$> b = fmap a b` 
-- We can also remove any parens except for those of the tuple
edgeOf p = fmap (poetryLockPackageName p,) . fmap PackageName . Map.keys $ poetryLockPackageDependencies p

-- Now it's more clear that we're `fmap`-ping twice
-- Since our functor is `List`, using `map` is equivalent here, let's try it
edgeOf p = map (poetryLockPackageName p,) . map PackageName . Map.keys $ poetryLockPackageDependencies p

-- Since map iterates over the list, and we call map twice, it will iterate twice.
-- Functor law #1 is (verbatim): `fmap f . fmap g = fmap (f . g)`
-- Since `fmap = map` for List, that means `map f . map g = map (f . g)`
-- Since we aren't calling map twice, it won't iterate twice
-- Let's rewrite to use this knowledge
edgeOf p = map ((poetryLockPackageName p,) . PackageName) . Map.keys $ poetryLockPackageDependencies p

-- Our first argument to `map` is a little verbose, let's make it more semantically clear with a name
edgeOf p = map tuplify . Map.keys $ poetryLockPackageDependencies p
  where
    tuplify = (poetryLockPackageName p,) . PackageName

-- For tuplify, we just copied the original implementation, but we can make it more clear now
edgeOf p = map tuplify . Map.keys $ poetryLockPackageDependencies p
  where
    tuplify x = (poetryLockPackageName p, PackageName x)
```

That was the exact logic I followed, with exact steps.

> Why use `map` when `fmap` works for any functor?

In short, clarity.  It's more clear to say "This is a list, we're mapping over the list", as opposed to, "we have _some_ functor and we want to apply this pure transformation to it".  When you know it's a list, tell the reader it's a list.  When you don't care about what functor you're applying your `a -> b` on, don't burden the reader with the knowledge, and just use `fmap`.

Also, not every functor will have a clear function like `map` for its `fmap` implementation.  `Maybe` implements fmap as a raw pattern match, rather than calling another function that's publicly exported.

Finally, `map` is a well-recognized function in multiple languages, so it gets a little bit of special treatment.  Using common terms for common concepts is a great way to communicate, and strong, clear communication is a big part of readability.

Reply:
This is absolutely great - Thanks Wes.